test/EndToEndTests/Common/Microsoft.OData.Client.E2E.TestCommon/Common/CustomHttpClientRequestMessage.cs

```diff
@@ -0,0 +1,127 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="CustomHttpClientRequestMessage.cs" company=".NET Foundation">
+//      Copyright (c) .NET Foundation and Contributors. All rights reserved.
+//      See License.txt in the project root for license information.
+// </copyright>
+//------------------------------------------------------------------------------
+
+namespace Microsoft.OData.Client.E2E.TestCommon.Common
+{
+    /// <summary>
+    /// An implementation of IODataRequestMessageAsync that uses an HttpRequestMessage under the covers.
+    /// In ODataLibrary, a message is an abstraction which consists of stream and header interfaces that hides the details of stream-reading/writing.
+    /// </summary>
+    public class CustomHttpClientRequestMessage : IODataRequestMessageAsync, IServiceCollectionProvider

```

Comment:
Why does the class need to implement `IServiceCollectionProvider`?

Reply:
In this instance it doesn't.. but this is a common class that will be used by several other tests.. and if there are tests that may require to pass custom services to the `ODataMessageReader` or `ODataMessageReader` through DI then this class needs to implement this interface. I know this a design flaw in our library and there are issues that have been raised before to address this. Like this one https://github.com/OData/odata.net/issues/2907. I think I'll leave it just to show the current state of how things are done.